lib/stages/visSharedMemWriter.cpp

```diff
@@ -0,0 +1,389 @@
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+
+
+REGISTER_KOTEKAN_STAGE(visSharedMemWriter);
+
+visSharedMemWriter::visSharedMemWriter(Config& config, const std::string& unique_name, bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&visSharedMemWriter::main_thread, this)) {
+
+        // Fetch any simple configuration
+        root_path = config.get_default<std::string>(unique_name, "root_path", "/dev/shm/");
+        sem_name = config.get_default<std::string>(unique_name, "sem_name", "kotekan");
+        fname_buf = config.get_default<std::string>(unique_name, "fname_buf", "calBuffer");
+        ntime = config.get_default<size_t>(unique_name, "nsamples", 512);
+
+        // Setup the input vector
+        in_buf = get_buffer("in_buf");
+        register_consumer(in_buf, unique_name.c_str());
+
+        // Check if any of the old buffer files exist
+        // Remove them, if they do
+        DEBUG("Checking for and removing old buffer files...");
+        check_remove(root_path + "sem." + sem_name);
+        check_remove(root_path + fname_buf);
+
+
+}
+
+visSharedMemWriter::~visSharedMemWriter() {
+    // make sure to unlink the semaphore and unmap the mappings
+    // We are setting num_writes to 0 in the structured data,
+    // to communicate to readers that the ring buffer is not being written to
+    num_writes = 0;
+    memcpy(structured_data_addr, &num_writes, sizeof(num_writes));
+}
+
+uint8_t* visSharedMemWriter::assign_memory(std::string shm_name, size_t shm_size) {
+    // takes the name of a shared memory address, opens a shared memory of the provided size, and maps the memory to a uint8_t* address pointer
+
+        uint8_t* addr;
+
+        int fd = shm_open(shm_name.c_str(), (O_CREAT | O_RDWR), (S_IRUSR | S_IWUSR));
+
+        if (fd == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Cannot open shared memory named {:s}: {:s}"), shm_name, strerror(errno)));
+        }
+
+        // Resize object to hold buffer
+        if (ftruncate(fd, shm_size) == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to expand shared memory named {:s}: {:s}"), shm_name, strerror(errno)));
+        }
+        INFO("Resized to {} bytes\n", (long) shm_size);
+
+        addr = (uint8_t*) mmap(nullptr, shm_size, (PROT_READ | PROT_WRITE), MAP_SHARED, fd, 0);
+        if (addr == MAP_FAILED) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to map shm {:s} to memory: {:s}."), shm_name, strerror(errno)));
+        }
+
+        // fd is no longer needed
+        if (close(fd) == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to close file descriptor for shm {:s}: {:s}."), shm_name, strerror(errno)));
+        }
+
+        return addr;
+}
+
+bool visSharedMemWriter::add_sample(const visFrameView& frame, time_ctype t, uint32_t freq_ind) {
```

Comment:
the mix of `return` and `else` in this method confused me for a good 20 minutes :spaghetti: